File: aiot_ws/src/simple_ros_cpp/src/service_server.cpp
#include "rclcpp/rclcpp.hpp"
#include "std_srvs/srv/set_bool.hpp"
#include <chrono>
#include <iostream>

using namespace std;
using namespace std::chrono_literals;

class ServiceServer : public rclcpp::Node
{
public:
    ServiceServer()
        : Node("service_server")
    {
        _server = create_service<std_srvs::srv::SetBool>("setBool", std::bind(&ServiceServer::service_callback, this, std::placeholders::_1, std::placeholders::_2));
    }

private:
    typedef std::shared_ptr<std_srvs::srv::SetBool::Request> BRequest;
    typedef std::shared_ptr<std_srvs::srv::SetBool::Response> BResponse;
    rclcpp::Service<std_srvs::srv::SetBool>::SharedPtr _server;
    void service_callback(const BRequest request, BResponse response)
    {
        std::string str = request->data ? std::string("true") : std::string("false");
        RCLCPP_INFO(get_logger(), str.c_str());
        response->message = "sucess";
        response->success = true;
    }
};

int main()
{
    rclcpp::init(0, nullptr);
    auto node = std::make_shared<ServiceServer>();
    rclcpp::spin(node);
    rclcpp::shutdown();
    return 0;
}

File: aiot_ws/src/simple_ros/simple_ros/service_server.py
import time

import rclpy
from rclpy.node import Node
from std_srvs.srv import SetBool


class Service_server(Node):
    def __init__(self):
        super().__init__("service_server")
        self.create_service(SetBool, "setBool", self.setBool_callback)
        self.bool = bool()
        self.cnt = 0

    def setBool_callback(self, request: SetBool.Request, response: SetBool.Response):
        self.get_logger().info(f"{self.cnt}번째 요청 처리")
        self.get_logger().info(f"현재 세팅된 불 값 : {self.bool}")
        self.get_logger().info(f"변경 요청한 불 값 : {request.data}")
        if request.data != self.bool:
            self.bool = not self.bool
            response.success = True
            response.message = f"{self.cnt}번째 요청 {self.bool} setting success"
        else:
            response.success = False
            response.message = f"{self.cnt}번째 요청 {self.bool} setting fail"
        time.sleep(5)
        return response


def main():
    rclpy.init()
    node = Service_server()
    try:
        rclpy.spin(node)
    except KeyboardInterrupt:
        node.destroy_node()


if __name__ == "__main__":
    main()
